// app/components/forecast.jsx
var React = require('react'),
    ForecastStore = require('app/stores/forecast'),
    StateMixin = require('react-router').State,
    StoreMixin = require('fluxible-app').StoreMixin,
    Hour = require('app/components/hour');

module.exports = React.createClass({
    mixins: [StoreMixin, StateMixin],

    statics: {
        storeListeners: {
            onChange: [ForecastStore]
        }
    },

    getInitialState: function () {
        var day = (this.getPath() === '/') ? 'today' : this.getParams().day;

        return {
            day: day,
            depart: parseInt(this.getQuery().depart, 10) || 8,
            return: parseInt(this.getQuery().return, 10) || 18,
            forecast: this.getStore(ForecastStore).getForecast(day)
        };
    },

    onChange: function () {
        this.setState({
            forecast: this.getStore(ForecastStore).getForecast()
        });
    },

    render: function () {
        return (
            <form className='forecast' method='get' action={ '/forecasts/' + this.state.day }>
                <Hour
                    forecast={ this.state.forecast.hourly.data }
                    defaultHour={ this.state.depart }
                    name='depart'
                    dawn={ this.state.forecast.daily.data[0].sunriseTime }
                    sunset={ this.state.forecast.daily.data[0].sunsetTime } />
                <Hour
                    forecast={ this.state.forecast.hourly.data }
                    defaultHour={ this.state.return }
                    name='return'
                    dawn={ this.state.forecast.daily.data[0].sunriseTime }
                    sunset={ this.state.forecast.daily.data[0].sunsetTime }
                    right={ true } />
                <input type='submit' value='Update forecast' className='updateButton' />
            </form>
        );
    }
});
