// app/components/forecast.jsx
var React = require('react'),
    ForecastStore = require('app/stores/forecast'),
    StateMixin = require('react-router').State,
    StoreMixin = require('fluxible-app').StoreMixin,
    Hour = require('app/components/hour');

module.exports = React.createClass({
    mixins: [StoreMixin, StateMixin],

    statics: {
        storeListeners: {
            onChange: [ForecastStore]
        }
    },

    getInitialState: function () {
        var day = (this.getPath() === '/') ? 'today' : this.getParams().day;

        return {
            day: day,
            forecast: this.getStore(ForecastStore).getForecast(day)
        };
    },

    onChange: function () {
        this.setState({
            forecast: this.getStore(ForecastStore).getForecast()
        });
    },

    render: function () {
        return (
            <main className='forecast'>
                <Hour
                    forecast={ this.state.forecast.hourly.data }
                    defaultTime={ 8 }
                    dawn={ this.state.forecast.daily.data[0].sunriseTime }
                    sunset={ this.state.forecast.daily.data[0].sunsetTime } />
                <Hour
                    forecast={ this.state.forecast.hourly.data }
                    defaultTime={ 18 }
                    dawn={ this.state.forecast.daily.data[0].sunriseTime }
                    sunset={ this.state.forecast.daily.data[0].sunsetTime }
                    right={ true } />
            </main>
        );
    }
});
