// app/components/forecasts/hour.jsx
var React = require('react'),
    TimePicker = require('app/components/forecasts/time-picker');

module.exports = React.createClass({
    propTypes: {
        forecast: React.PropTypes.array.isRequired,
        defaultHour: React.PropTypes.number.isRequired,
        name: React.PropTypes.string.isRequired,
        dawn: React.PropTypes.number.isRequired,
        sunset: React.PropTypes.number.isRequired,
        right: React.PropTypes.bool
    },

    getDefaultProps: function () {
        return {
            right: false
        };
    },

    getInitialState: function () {
        return {
            hour: this.props.defaultHour
        };
    },

    withinAnHourOf: function (point) {
        return (this.props.forecast[this.state.hour].time < this.props[point] + 1800 &&
            this.props.forecast[this.state.hour].time > this.props[point] - 1800);
    },

    isNight: function () {
        return this.props.forecast[this.state.hour].time < this.props.dawn - 1800 ||
            this.props.forecast[this.state.hour].time > this.props.sunset + 1800;
    },

    render: function () {
        var classes = 'hour' + 
            ((this.props.right) ? ' right' : '') +
            ((this.withinAnHourOf('dawn')) ? ' dawn' : '') +
            ((this.withinAnHourOf('sunset')) ? ' sunset' : '') + 
            ((this.isNight()) ? ' night' : '');

        return (
            <div className={ classes }>
                <p className='summary'>
                    <TimePicker hour={ this.state.hour } name={ this.props.name } />
                    <span className='temperature icon-sunny'>
                        { Math.ceil(this.props.forecast[this.state.hour].temperature) }&deg;
                    </span>
                </p>
            </div>
        );
    }
});
